File: 백준/Gold/11657. 타임머신/타임머신.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main {
	private static class Node {
		int start;
		int end;
		int cost;

		Node(int start, int end, int cost) {
			this.start = start;
			this.end = end;
			this.cost = cost;
		}

	}

	static boolean flag = false;
	static long[] dist;
	static Node[] map;
	static int INF = 987654321;
	static int N,M;

	public static void main(String[] args) throws IOException {
		// 음의 가중치가 있는 문제
		// 벨만-포드 알고리즘
		// 모든 간선에 대해 시작 노드에서 해당 간선의 시작점까지 현재 거리와 간선의 가중치의 합을 구함
		// 이 합이 간선의 끝점까지의 현재 거리보다 작다면, 간선의 끝점까지의 거리를 업데이트
		// 이 작업을 노드의 수-1 번 반복
		// 모든 간선을 다시 한번 검사해서 위의 업데이트가 또 반복된다면 이는 음의 사이클이 존재함을 의미
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		// 경로 저장
		dist = new long[N + 1];
		Arrays.fill(dist, INF);
		map = new Node[M];

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int start = Integer.parseInt(st.nextToken());
			int end = Integer.parseInt(st.nextToken());
			int cost = Integer.parseInt(st.nextToken());

			map[i] = new Node(start, end, cost);
		}
		
		if(bellmanford(1)) { // 음수 순환 존재하면 -1 출력 
			System.out.println(-1);
		}
		else {
			// 1번 노드를 제외한 다른 모든 노드로 가기 위한 최단거리 출력 
			for(int i=2; i<N+1; i++) {
				if(dist[i] == INF) {// 도달할 수 없으면 -1 
					System.out.println("-1");
				}
				else { // 최단 거리 출력 
					System.out.println(dist[i]);
				}
			}
		}

	}

	static boolean bellmanford(int start){
		dist[start] = 0;
		for(int i=1;i<N+1;i++) {
			for(int j=0;j<M;j++) {
				int cur = map[j].start;
				int next = map[j].end;
				int cost = map[j].cost;
				
				if(dist[cur]==INF) continue;
				
				if(dist[next] > dist[cur]+cost) {
					dist[next] =dist[cur] + cost;
					if(i==N) {
						return true;
					}
				}
			}
		}
		return false;
	}
}

File: 백준/Gold/11657. 타임머신/타임머신.cc
#include <iostream>
#include <vector>
using namespace std;
#define INF 987654321

// 1번에서 다른 도시로 가는 최소 시간
// 음수 사이클이 있음 -> 벨만 포드
// 최소 시간을 구하고 한번 더 돌려야 한다(시간이 감소하는지 확인하기 위함)


int N, M;
vector<pair<int,pair<int, int>>> roadList;
long long dist[501];
bool flag = false;


bool bellmanFord(int start)
{
	dist[start] = 0;
	// N개의 도시를 다 돌라면? N-1번 순환
	// 즉 음수 사이클이 없다면 N-1번 도는 동안 모든 경로의 최적화가 마무리
	// 그런데 N번째 순환에서 경로 최적화가 발생한다? 음수 사이클이 있어 무한 경로가 발생한다는 뜻
	for (int i = 0;i <= N;i++) 
	{
		for (auto j = 0;j < roadList.size();j++) // 반복자로 처음 주소부터 받아오기
		{
			int cur = roadList[j].first;
			int next = roadList[j].second.first;
			int cost = roadList[j].second.second;

			if (dist[cur] == INF) continue; // 아직 현재 경로가 없는 부분은 지나간다.
			if (dist[next] > dist[cur] + cost)
			{
				dist[next] = dist[cur] + cost;
				if (i == N) // N번째 순환에서 경로의 최적화 발생
					return true;
			}
		}
	}

	return false;
}

int main()
{

	cin >> N >> M;

	for (int i = 1;i <= N;i++)
	{
		dist[i] = INF;
	}
	for (int i = 0;i < M;i++)
	{
		int st, ed, ti;
		cin >> st >> ed >> ti;
		roadList.push_back(make_pair(st,make_pair(ed,ti))); // make_pair 로 pair 만들고 push_back으로 뒤로 추가
	}

	if (bellmanFord(1))
		cout << "-1"; // true 반환 시 음수 사이클이 있다
	else
	{
		for (int i = 2;i <= N;i++)
		{
			if (dist[i] == INF)
				cout << -1 << '\n';
			else
				cout << dist[i] << '\n';
		}
	}
	return 0;
}